RockSniffer/RPC/AlbumArtResolver.cs

```diff
@@ -0,0 +1,110 @@
+﻿using Newtonsoft.Json;
+using RockSnifferLib.Sniffing;
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Dynamic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace RockSniffer.RPC
+{
+    class AppleResult
+    {
+        public string artistName;
+        public string collectionName;
+        public string trackName;
+        public string artworkUrl100;
+    }
+
+    class AppleRequestResponse {
+        public int resultCount;
+        public AppleResult[] results;
+    }
+
+    public class AlbumArtResolver
+    {
+        private HttpClient httpClient = new HttpClient();
+        private Dictionary<string, (string URL, string DisplayText)> cache = new Dictionary<string, (string URL, string DisplayText)>();
+
+        /// <summary>
+        /// Attempt to obtain album cover
+        /// </summary>
+        /// <param name="songInfo">Details about the given song</param>
+        public (string? URL, string DisplayText) Get(SongDetails songInfo)
+        {
+            string key = $"{songInfo.artistName}|{songInfo.albumName}";
+
+            // Use double-checked locking pattern, since there are multiple threads
+            if (!cache.Keys.Contains(key))
```

Comment:
If multiple threads are accessing the cache dictionary, consider using [ConcurrentDictionary](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-6.0)

Reply:
I think ConcurrentDictionary wouldn't help in this case: the issue is between checking the key doesn't exist and inserting it in, another thread could have inserted.

Reply:
I think it would, it's designed for thread safety, and handles locking internally. I'd prefer using the builtin class over manual locking, if possible.

Is there a reason why you check the key through `Keys.Contains` instead of `ContainsKey`?

Reply:
Well, I looked into it, and in different circumstances you could use something like `AddOrUpdate(key, (key) => CalculateValue(), (key, val) => val)`, but given that it is not known whether there even will be a result, I don't think there is a good way to do it.

Reply:
You shouldn't need to use the delegate functions, just `cache.ContainsKey` and `cache.Add`? Or am I missing something?

Reply:
Well, I believe so. Think about situation with this code and two threads:
```csharp
if (!cache.ContainsKey(key))
  cache.Add(key, value);
```
If both threads pass the condition, both will attempt to add to the cache and the second one will cause Exception, regardless of how the cache is implemented.

Reply:
Ah, I see your point now, even when using `ConcurrentCollection.TryAdd` method which handles this case, both threads would do the work of fetching.
Thank you for clarifying.